src/topology/metric_space/isometric_smul.lean

```diff
@@ -0,0 +1,370 @@
+  c • ball x r = ball (c • x) r :=
+(isometric.const_smul c).image_ball _ _
+
+@[simp, to_additive] lemma preimage_smul_ball (c : G) (x : X) (r : ℝ) :
+  ((•) c) ⁻¹' ball x r = ball (c⁻¹ • x) r :=
+by rw [preimage_smul, smul_ball]
+
+@[simp, to_additive] lemma smul_closed_ball (c : G) (x : X) (r : ℝ) :
+  c • closed_ball x r = closed_ball (c • x) r :=
+(isometric.const_smul c).image_closed_ball _ _
+
+@[simp, to_additive] lemma preimage_smul_closed_ball (c : G) (x : X) (r : ℝ) :
+  ((•) c) ⁻¹' closed_ball x r = closed_ball (c⁻¹ • x) r :=
+by rw [preimage_smul, smul_closed_ball]
+
+@[simp, to_additive] lemma smul_sphere (c : G) (x : X) (r : ℝ) :
+  c • sphere x r = sphere (c • x) r :=
+(isometric.const_smul c).image_sphere _ _
+
+@[simp, to_additive] lemma preimage_smul_sphere (c : G) (x : X) (r : ℝ) :
+  ((•) c) ⁻¹' sphere x r = sphere (c⁻¹ • x) r :=
+by rw [preimage_smul, smul_sphere]
+
+variables [pseudo_metric_space G]
+
+@[simp, to_additive]
+lemma preimage_mul_left_ball [has_isometric_smul G G] (a b : G) (r : ℝ) :
+  ((*) a) ⁻¹' ball b r = ball (a⁻¹ * b) r :=
+preimage_smul_ball a b r
+
+@[simp, to_additive]
+lemma preimage_mul_right_ball [has_isometric_smul Gᵐᵒᵖ G] (a b : G) (r : ℝ) :
+  (λ x, x * a) ⁻¹' ball b r = ball (b / a) r :=
+by { rw div_eq_mul_inv, exact preimage_smul_ball (mul_opposite.op a) b r }
+
+@[simp, to_additive]
+lemma preimage_mul_left_closed_ball [has_isometric_smul G G] (a b : G) (r : ℝ) :
+  ((*) a) ⁻¹' closed_ball b r = closed_ball (a⁻¹ * b) r :=
+preimage_smul_closed_ball a b r
+
+@[simp, to_additive]
+lemma preimage_mul_right_closed_ball [has_isometric_smul Gᵐᵒᵖ G] (a b : G) (r : ℝ) :
+  (λ x, x * a) ⁻¹' closed_ball b r = closed_ball (b / a) r :=
+by { rw div_eq_mul_inv, exact preimage_smul_closed_ball (mul_opposite.op a) b r }
+
+end metric
+
+section instances
+
+variables {Y : Type*} [pseudo_emetric_space X] [pseudo_emetric_space Y] [has_smul M X]
+  [has_isometric_smul M X]
+
+@[to_additive] instance [has_smul M Y] [has_isometric_smul M Y] :
+  has_isometric_smul M (X × Y) :=
+⟨λ c, (isometry_smul X c).prod_map (isometry_smul Y c)⟩
+
+@[to_additive] instance prod.has_isometric_smul' {N}
+  [has_mul M] [pseudo_emetric_space M] [has_isometric_smul M M]
+  [has_mul N] [pseudo_emetric_space N] [has_isometric_smul N N] :
+  has_isometric_smul (M × N) (M × N) :=
+⟨λ c, (isometry_smul M c.1).prod_map (isometry_smul N c.2)⟩
+
+@[to_additive] instance prod.has_isometric_smul'' {N}
+  [has_mul M] [pseudo_emetric_space M] [has_isometric_smul Mᵐᵒᵖ M]
+  [has_mul N] [pseudo_emetric_space N] [has_isometric_smul Nᵐᵒᵖ N] :
+  has_isometric_smul (M × N)ᵐᵒᵖ (M × N) :=
+⟨λ c, (isometry_mul_right c.unop.1).prod_map (isometry_mul_right c.unop.2)⟩
+
+@[to_additive] instance units.has_isometric_smul [monoid M] : has_isometric_smul Mˣ X :=
+⟨λ c, by convert isometry_smul X (c : M)⟩
+
+@[to_additive] instance : has_isometric_smul M Xᵐᵒᵖ :=
+⟨λ c x y, by simpa only using edist_smul_left c x.unop y.unop⟩
```

Comment:
One other good instances would be
```lean
@[to_additive] instance [is_central_scalar M X] : has_isometric_smul Mᵐᵒᵖ X :=
⟨λ c x y, sorry⟩
```

Reply:
(this was actually the one I was thinking of before)

Reply:
Will this create a problem with the way we handle `normed_group`s? Our `normed_group`s have right-invariant distances and deduce left-invariance from commutativity.

Reply:
This shouldn't cause any diamond probems because the instance carries no data.

It's possible that my instance makes the instances you're thinking of redundant.

Reply:
The problem is that for `normed_group`s (with the current definition), the implication goes in the other direction. Though this shouldn't create a loop because of the `[normed_group _]` dependency.